Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,16 +427,101 @@ Further information:
         }
     }
 
+    private func run(_ task: UnderlyingTask) {
+        /* for macOS: in case any calls we make to Foundation put objects into an autoreleasepool */
+        withAutoReleasePool {
+            switch task {
+            case .function(let function):
+                function()
+            #if compiler(>=5.9)
+            case .unownedJob(let erasedUnownedJob):
+                if #available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *) {
+                    erasedUnownedJob.unownedJob.runSynchronously(on: self.asUnownedSerialExecutor())
+                } else {
+                    fatalError("Tried to run an UnownedJob without runtime support")
+                }
+            #endif
+            }
+        }
+    }
+    
+    private func runLoop() -> NIODeadline? {
+        // We need to ensure we process all tasks, even if a task added another task again
+        while true {
+            // TODO: Better locking
+            let nextReadyDeadline = self._tasksLock.withLock { () -> NIODeadline? in
+                var moreImmediateTasksToConsider = !self._immediateTasks.isEmpty
+                var moreScheduledTasksToConsider = !self._scheduledTasks.isEmpty
+                
+                guard moreImmediateTasksToConsider || moreScheduledTasksToConsider else {
+                    // We will not continue to loop here. We need to be woken if a new task is enqueued.
+                    self._pendingTaskPop = false
+                    
+                    // Reset nextReadyDeadline to nil which means we will do a blocking select.
+                    return nil
+                }
+                
+                // We only fetch the time one time as this may be expensive and is generally good enough as if we miss anything we will just do a non-blocking select again anyway.
+                let now: NIODeadline = .now()
+                var nextScheduledTaskDeadline = now
+
+                while self.tasksCopy.count < self.tasksCopy.capacity &&
```

Comment:
@gmilos why do we consider `self.tasksCopy.capacity`? That's essentially a random value

Reply:
This is how this code always worked (except I span off this block to a separate function), see:
https://github.com/apple/swift-nio/pull/2645/files#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faL503

The value isn't random, it's set to `4096`, and never changed:
https://github.com/apple/swift-nio/pull/2645/files#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faR188

I infer we wanted to get batch of tasks out, under the lock. And use `taskCopy` as completely "static" memory block to shuffle tasks out from the `_scheduledTasks` heap (and now also `_immediateTasks` circular buffer).

I would appreciate review of the logic here though. It's a bit fiddly. Especially around when is the `_pendingTaskPop` meant to be set and what value should `nextScheduledTaskDeadline` assume.

Reply:
oh dear, I had no idea :). That's super duper fragile, capacity is _at least_ what `reserveCapacity` did, unless we ever call `removeAll` forgetting to pass `keepingCapacity: true`. Would you mind fixing this long-standing bug and just comparing it to a constant? As in `self.tasksCopy.count < Self.maximumEventLoopTaskExecutionBatchSize` or something?

Reply:
fixed here: https://github.com/apple/swift-nio/pull/2645/commits/f76e65bb71f92bd35b532bbeb477d8ec1953a328

I happen to have picked a different name for the constant (before reading your comment here). Same idea though.